File: ios/Classes/Configuration/Preferences.swift
import Foundation
import iOSPhoneLib

internal func preferencesOf(_ dict: Dictionary<String, Any?>) -> Preferences {
    NSLog(String(describing: dict))

    return Preferences(
        useApplicationRingtone: dict["useApplicationProvidedRingtone"] as? Bool ?? false,
        includesCallsInRecents: dict["showCallsInNativeRecents"] as? Bool ?? true,
        supplementaryContacts: Set((dict["supplementaryContacts"] as? Array<Dictionary<String, String?>> ?? []).map {
            (item) in  SupplementaryContact(
                number: (item["number"] ?? "") ?? "",
                name: (item["name"] ?? "") ?? ""
            )
        }
        enableAdvancedLogging: dict["enableAdvancedLogging"] as? Bool ?? false,
    ))
}

internal func preferencesOf(_ json: String?) -> Preferences? {
    if (json == nil) { return nil }
    
    return preferencesOf(deserializeFromJSON(json!))
}

internal extension Preferences {
    func serialize() -> String {
        return serializeToJSON([
            "useApplicationRingtone": useApplicationRingtone,
            "showCallsInNativeRecents": includesCallsInRecents,
            "supplementaryContacts": supplementaryContacts.map { item in ["name": item.name, "number": item.number]},
        ])
    }
}

File: android/src/main/kotlin/org/openvoipalliance/flutterphonelib/configuration/Preferences.kt
package org.openvoipalliance.flutterphonelib.configuration

import org.openvoipalliance.androidphoneintegration.configuration.Preferences

fun preferencesOf(map: Map<String, Any?>) = object {
    val useApplicationProvidedRingtone: Boolean by map
    val enableAdvancedLogging: Boolean by map

    val preferences = Preferences(
        useApplicationProvidedRingtone,
        enableAdvancedLogging,
    )
}.preferences
